Sources/SwiftSyntax/Identifier.swift

```diff
@@ -12,23 +12,38 @@
 
 /// A canonicalized representation of an identifier that strips away backticks.
 public struct Identifier: Equatable, Hashable, Sendable {
-  /// The sanitized `text` of a token.
+  /// The sanitized name of the identifier.
   public var name: String {
     String(syntaxText: raw.name)
   }
 
   @_spi(RawSyntax)
   public let raw: RawIdentifier
-
-  private let arena: SyntaxArenaRef
+  let arena: SyntaxArenaRef?
 
   public init?(_ token: TokenSyntax) {
     guard case .identifier = token.tokenKind else {
       return nil
     }
 
     self.raw = RawIdentifier(token.tokenView)
-    self.arena = token.tokenView.raw.arenaReference
+    self.arena = token.raw.arenaReference
+  }
+
+  public init(_ staticString: StaticString) {
+    self.raw = RawIdentifier(staticString)
+    self.arena = nil
+  }
+
+  public static func == (lhs: Self, rhs: Self) -> Bool {
+    lhs.name == rhs.name
+  }
+}
+
+extension Identifier {
+  @_spi(Testing) public init(anyToken token: TokenSyntax) {
+    self.raw = RawIdentifier(token.tokenView.rawText)
+    self.arena = token.raw.arenaReference
```

Comment:
I’m not a fan of this initializer because it deliberately discards the difference between keywords and identifiers in a very subtle way. I assume that we only need it for `self` and `Self`, right? Since these behave identifier-like, what do you think about changing the existing `init?(_ token: TokenSyntax)` to initialize itself with `self` and `Self` if the passed in token is `self` or `Self`?

I haven’t thought it through but this might also simplify the name lookup logic because you no longer need to differentiate between the `self` keyword and `self` identifier`. Maybe `LookupName.self` and `LookupName.Self` won’t be needed anymore.

Reply:
I think this is a good idea indeed. Not being able to create an `Identifier` from `self` and `Self` caused a lot of trouble that now could be handled out of the box.